---
title: Exercises 2
format:
  live-html:
    pyodide:
      cell-options:
        autorun: false
        completion: true
---

## Find the most common words

Download a text file, like [Alice in wonderland](https://www.gutenberg.org/ebooks/11.txt.utf-8) from [Project Gutenberg](https://www.gutenberg.org/), count the number of time each word appears and print the 50 most common words.

Take into account that the Alice in wonderland text really only starts after the line "CHAPTER I." and ends with the line "THE END".

```{pyodide}
#| exercise: count_words

```

::: { .solution exercise="count_words" }
::: { .callout-tip collapse="false"}

A solution using a dictionary to keep the counts.

```{pyodide}
file_name = "11.txt.utf-8"
file = open(file_name, 'rt')

in_alice = False
counts = {}
for line in file:
    line = line.strip()
    if line == "CHAPTER I.":
        in_alice = True
    elif line == "THE END":
        in_alice = False
    if not in_alice:
        continue
    
    line = line.lower()
    for character in [":", ",", ".", "-", "(", ")", "—", "_", "-", "“", "”", "!", "?"]:
        line = line.replace(character, " ")
    for word in line.split():
        if word in counts:
            counts[word] += 1
        else:
            counts[word] = 1

words_and_counts = list(reversed(sorted(counts.items(), key=lambda word_count: word_count[1])))[:50]
for word, count in words_and_counts:
    print(word, count)
```

A solution using the [Counter](https://docs.python.org/3/library/collections.html#collections.Counter) class from the [collections](https://docs.python.org/3/library/collections.html) module.

```{pyodide}
from collections import Counter

file_name = "11.txt.utf-8"
file = open(file_name, 'rt')

in_alice = False
counts = Counter()
for line in file:
    line = line.strip()
    if line == "CHAPTER I.":
        in_alice = True
    elif line == "THE END":
        in_alice = False
    if not in_alice:
        continue
    
    line = line.lower()
    for character in [":", ",", ".", "-", "(", ")", "—", "_", "-", "“", "”", "!", "?"]:
        line = line.replace(character, " ")
    words = line.split()
    counts.update(words)

for word, count in counts.most_common(50):
    print(word, count)
```
:::
:::

## Read the genetic code into a dictionary

The dictionary should have the codon as the key and the aminoacid name as the value.

```{pyodide}
#| exercise: genetic_code_dict
AA_TABLE = '''Codon aa_full_name aa_abreviation
TTT Phenylalanine F
TTC Phenylalanine F
TTA Leucine L
TTG Leucine L
TCT Serine S
TCC Serine S
TCA Serine S
TCG Serine S
TAT Tyrosine Y
TAC Tyrosine Y
TAA Stop X
TAG Stop  X
TGT Cysteine C
TGC Cysteine C
TGA Stop X
TGG Tryptophan W
CTT Leucine L
CTC Leucine L
CTA Leucine L
CTG Leucine L
CCT Proline P
CCC Proline P
CCA Proline P
CCG Proline P
CAT Histidine H
CAC Histidine H
CAA Glutamine Q
CAG Glutamine Q
CGT Arginine R
CGC Arginine R
CGA Arginine R
CGG Arginine R
ATT Isoleucine I
ATC Isoleucine I
ATA Isoleucine I
ATG Methionine M
ACT Threonine T
ACC Threonine T
ACA Threonine T
ACG Threonine T
AAT Asparagine N
AAC Asparagine N
AAA Lysine K
AAG Lysine K
AGT Serine S
AGC Serine S
AGA Arginine R
AGG Arginine R
GTT Valine V
GTC Valine V
GTA Valine V
GTG Valine V
GCT Alanine A
GCC Alanine A
GCA Alanine A
GCG Alanine A
GAT Aspartate D
GAC Aspartate D
GAA Glutamate E
GAG Glutamate E
GGT Glycine G
GGC Glycine G
GGA Glycine G
GGG Glycine G
'''
```
::: { .hint exercise="genetic_code_dict"}
::: { .callout-note collapse="false"}

splitlines

:::
:::
::: { .solution exercise="genetic_code_dict"}
::: { .callout-tip collapse="false"}
```{pyodide}
AA_TABLE = '''Codon aa_full_name aa_abreviation
TTT Phenylalanine F
TTC Phenylalanine F
TTA Leucine L
TTG Leucine L
TCT Serine S
TCC Serine S
TCA Serine S
TCG Serine S
TAT Tyrosine Y
TAC Tyrosine Y
TAA Stop X
TAG Stop  X
TGT Cysteine C
TGC Cysteine C
TGA Stop X
TGG Tryptophan W
CTT Leucine L
CTC Leucine L
CTA Leucine L
CTG Leucine L
CCT Proline P
CCC Proline P
CCA Proline P
CCG Proline P
CAT Histidine H
CAC Histidine H
CAA Glutamine Q
CAG Glutamine Q
CGT Arginine R
CGC Arginine R
CGA Arginine R
CGG Arginine R
ATT Isoleucine I
ATC Isoleucine I
ATA Isoleucine I
ATG Methionine M
ACT Threonine T
ACC Threonine T
ACA Threonine T
ACG Threonine T
AAT Asparagine N
AAC Asparagine N
AAA Lysine K
AAG Lysine K
AGT Serine S
AGC Serine S
AGA Arginine R
AGG Arginine R
GTT Valine V
GTC Valine V
GTA Valine V
GTG Valine V
GCT Alanine A
GCC Alanine A
GCA Alanine A
GCG Alanine A
GAT Aspartate D
GAC Aspartate D
GAA Glutamate E
GAG Glutamate E
GGT Glycine G
GGC Glycine G
GGA Glycine G
GGG Glycine G
'''

genetic_code = {}
first_line = True
for line in AA_TABLE.splitlines():
    if first_line:
        first_line = False
        continue
    line = line.strip()
    if not line:
        continue
    codon, long_name, short_name = line.split()
    genetic_code [codon] = short_name
print(genetic_code)
```
:::
:::

## Write a program that translate a coding sequence into a protein

```{pyodide}
#| exercise: translate_protein

dna_coding_seq = 'atgtggacttga'
prot_seq = translate(dna_coding_seq)
```

::: { .hint exercise="translate_protein"}
::: { .callout-note collapse="false"}

You can get the position of every codon using the [range](https://docs.python.org/3/library/functions.html#func-range) function.
Remember that range has a step argument that can be set to 3.

:::
:::
::: { .solution exercise="translate_protein" }
::: { .callout-tip collapse="false"}
```{pyodide}
AA_TABLE = '''Codon aa_full_name aa_abreviation
TTT Phenylalanine F
TTC Phenylalanine F
TTA Leucine L
TTG Leucine L
TCT Serine S
TCC Serine S
TCA Serine S
TCG Serine S
TAT Tyrosine Y
TAC Tyrosine Y
TAA Stop X
TAG Stop  X
TGT Cysteine C
TGC Cysteine C
TGA Stop X
TGG Tryptophan W
CTT Leucine L
CTC Leucine L
CTA Leucine L
CTG Leucine L
CCT Proline P
CCC Proline P
CCA Proline P
CCG Proline P
CAT Histidine H
CAC Histidine H
CAA Glutamine Q
CAG Glutamine Q
CGT Arginine R
CGC Arginine R
CGA Arginine R
CGG Arginine R
ATT Isoleucine I
ATC Isoleucine I
ATA Isoleucine I
ATG Methionine M
ACT Threonine T
ACC Threonine T
ACA Threonine T
ACG Threonine T
AAT Asparagine N
AAC Asparagine N
AAA Lysine K
AAG Lysine K
AGT Serine S
AGC Serine S
AGA Arginine R
AGG Arginine R
GTT Valine V
GTC Valine V
GTA Valine V
GTG Valine V
GCT Alanine A
GCC Alanine A
GCA Alanine A
GCG Alanine A
GAT Aspartate D
GAC Aspartate D
GAA Glutamate E
GAG Glutamate E
GGT Glycine G
GGC Glycine G
GGA Glycine G
GGG Glycine G
'''

def create_genetic_code():
    genetic_code = {}
    first_line = True
    for line in AA_TABLE.splitlines():
        if first_line:
            first_line = False
            continue
        line = line.strip()
        if not line:
            continue
        codon, long_name, short_name = line.split()
        genetic_code [codon] = short_name
    return genetic_code

def translate(dna_seq):
    genetic_code = create_genetic_code()
    aminoacids = []
    for codon_start in range(0, len(dna_seq), 3):
        codon = dna_seq[codon_start: codon_start + 3]
        aminoacid = genetic_code[codon.upper()]
        aminoacids.append(aminoacid)
    return ''.join(aminoacids)
    
dna_coding_seq = 'atgtggacttga'
prot_seq = translate(dna_coding_seq)
print(prot_seq)
```
:::
:::

## Gene expression

We have done a study in which several patients have been given or not given (the placebo group) a drug.
We have measured the expression of three genes and now we want to know which is the mean expression for each gene for the treated and placebo groups.
We have also recorded the patient sex and we want the mean expression of each gene for males and females and the combined effect of sex and treament, so male-treated, male-placebo, female-treated, and female-placebo.

```{pyodide}
#| exercise: grouped_gene_expression
patient_data = {'patient1': {'placebo': 'yes', 'sex': 'M', 'gene1': 7.31, 'gene2': 12.15, 'gene3': 13.29}, 'patient2': {'placebo': 'yes', 'sex': 'M', 'gene1': 2.08, 'gene2': 19.26, 'gene3': 17.61}, 'patient3': {'placebo': 'yes', 'sex': 'M', 'gene1': 1.77, 'gene2': 12.34, 'gene3': 7.82}, 'patient4': {'placebo': 'no', 'sex': 'M', 'gene1': 14.2, 'gene2': 19.28, 'gene3': 11.41}, 'patient5': {'placebo': 'no', 'sex': 'M', 'gene1': 9.05, 'gene2': 15.12, 'gene3': 8.96}, 'patient6': {'placebo': 'yes', 'sex': 'M', 'gene1': 14.73, 'gene2': 14.64, 'gene3': 16.88}, 'patient7': {'placebo': 'no', 'sex': 'M', 'gene1': 8.0, 'gene2': 4.57, 'gene3': 8.82}, 'patient8': {'placebo': 'yes', 'sex': 'M', 'gene1': 16.46, 'gene2': 18.43, 'gene3': 8.69}, 'patient9': {'placebo': 'yes', 'sex': 'M', 'gene1': 9.07, 'gene2': 8.32, 'gene3': 6.8}, 'patient10': {'placebo': 'yes', 'sex': 'M', 'gene1': 1.59, 'gene2': 17.97, 'gene3': 10.85}, 'patient11': {'placebo': 'no', 'sex': 'M', 'gene1': 18.94, 'gene2': 12.81, 'gene3': 5.07}, 'patient12': {'placebo': 'yes', 'sex': 'F', 'gene1': 13.17, 'gene2': 15.11, 'gene3': 10.6}, 'patient13': {'placebo': 'no', 'sex': 'M', 'gene1': 18.91, 'gene2': 16.28, 'gene3': 9.74}, 'patient14': {'placebo': 'no', 'sex': 'M', 'gene1': 3.92, 'gene2': 12.25, 'gene3': 7.17}, 'patient15': {'placebo': 'yes', 'sex': 'M', 'gene1': 14.27, 'gene2': 9.18, 'gene3': 5.43}, 'patient16': {'placebo': 'yes', 'sex': 'M', 'gene1': 9.8, 'gene2': 8.58, 'gene3': 7.6}, 'patient17': {'placebo': 'yes', 'sex': 'F', 'gene1': 17.16, 'gene2': 2.61, 'gene3': 12.57}, 'patient18': {'placebo': 'no', 'sex': 'M', 'gene1': 10.5, 'gene2': 7.63, 'gene3': 9.88}, 'patient19': {'placebo': 'yes', 'sex': 'M', 'gene1': 4.98, 'gene2': 17.15, 'gene3': 18.29}, 'patient20': {'placebo': 'yes', 'sex': 'F', 'gene1': 4.51, 'gene2': 8.79, 'gene3': 10.68}, 'patient21': {'placebo': 'yes', 'sex': 'F', 'gene1': 8.16, 'gene2': 11.03, 'gene3': 6.67}, 'patient22': {'placebo': 'no', 'sex': 'M', 'gene1': 4.81, 'gene2': 15.38, 'gene3': 19.0}, 'patient23': {'placebo': 'no', 'sex': 'F', 'gene1': 9.06, 'gene2': 10.09, 'gene3': 10.23}, 'patient24': {'placebo': 'no', 'sex': 'F', 'gene1': 15.12, 'gene2': 16.36, 'gene3': 8.39}, 'patient25': {'placebo': 'no', 'sex': 'F', 'gene1': 2.61, 'gene2': 16.31, 'gene3': 15.65}, 'patient26': {'placebo': 'yes', 'sex': 'M', 'gene1': 13.9, 'gene2': 9.6, 'gene3': 4.46}, 'patient27': {'placebo': 'no', 'sex': 'M', 'gene1': 15.63, 'gene2': 3.93, 'gene3': 19.77}, 'patient28': {'placebo': 'no', 'sex': 'M', 'gene1': 15.45, 'gene2': 6.27, 'gene3': 4.81}, 'patient29': {'placebo': 'no', 'sex': 'M', 'gene1': 6.68, 'gene2': 16.39, 'gene3': 16.65}, 'patient30': {'placebo': 'yes', 'sex': 'M', 'gene1': 17.39, 'gene2': 5.87, 'gene3': 10.63}, 'patient31': {'placebo': 'no', 'sex': 'F', 'gene1': 5.72, 'gene2': 1.13, 'gene3': 15.77}, 'patient32': {'placebo': 'no', 'sex': 'M', 'gene1': 14.72, 'gene2': 11.7, 'gene3': 17.29}, 'patient33': {'placebo': 'no', 'sex': 'M', 'gene1': 15.42, 'gene2': 10.81, 'gene3': 17.73}, 'patient34': {'placebo': 'no', 'sex': 'F', 'gene1': 9.01, 'gene2': 7.65, 'gene3': 10.27}, 'patient35': {'placebo': 'yes', 'sex': 'M', 'gene1': 19.64, 'gene2': 3.51, 'gene3': 16.52}}
GENES = ("gene1", "gene2", "gene3")

def filter_patient_data(patient_data, desired_placebo=None, desired_sex=None):
    filtered_patients = []
    for patient in patient_data.values():
        patient_sex = patient['sex']
        if desired_sex is not None and patient_sex != desired_sex:
            continue
      
        patient_placebo = patient['placebo']
        if patient_placebo == 'yes':
            patient_placebo = True
        else:
            patient_placebo = False
        if desired_placebo is not None and patient_placebo != desired_placebo:
            continue

        filtered_patients.append(patient)
    return filtered_patients


def gather_expression_by_gene(patient_list):
    gene_expressions = {}
    for gene in GENES:
        gene_expressions[gene] = []

    for patient in patient_list:
        for gene in GENES:
            gene_expressions[gene].append(patient[gene])
    return gene_expressions


def calc_mean_expressions(expressions_per_gene):
    mean_expressions = {}
    for gene, values in expressions_per_gene.items():
        mean_expressions[gene] = sum(values) / len(values)
    return mean_expressions

def calc_mean_expressions_for_patients(patient_data, desired_placebo=None, desired_sex=None):
    filtered_patients = filter_patient_data(patient_data,
                                            desired_placebo=desired_placebo,
                                            desired_sex=desired_sex)
    expressions_per_gene = gather_expression_by_gene(filtered_patients)
    mean_expressions = calc_mean_expressions(expressions_per_gene)
    return mean_expressions


for sex in ("M", "F"):
    mean_expressions = calc_mean_expressions_for_patients(patient_data, desired_placebo=None, desired_sex=sex)
    print(f"sex {sex}")
    print(mean_expressions)

```

## Guess the number game

Write a program that chooses a random number between 1 and 10 and asks the user to guess it several times.
The program should finish once the user guesses correctly or once the number of available attemps has run out.

```{pyodide}
#| exercise: guess_random_number
import random

max_num_attemps = 3

max_number = 10
number_to_guess = random.randint(1, max_number)
```
::: { .hint exercise="guess_random_number"}
::: { .callout-note collapse="false"}

The function [randint](https://docs.python.org/3/library/random.html#random.randint) of the [random](https://docs.python.org/3/library/random.html) module creates random integers.

You can ask input from a user, like his guess, using the [input](https://docs.python.org/3/library/functions.html#input) function.
Remember that the input function always return a string.

:::
:::
::: { .solution exercise="guess_random_number" }
::: { .callout-tip collapse="false"}
```{pyodide}
import random

max_num_attemps = 3

max_number = 10
number_to_guess = random.randint(1, max_number)

guessed_correctly = False
for idx in range(max_num_attemps):
    guess = input(f"Attempt {idx + 1}, guess a number between 1 and {max_number}: ")
    guess = int(guess)
    if guess == number_to_guess:
        guessed_correctly = True
        break

if guessed_correctly:
    print(f"Congratulations you have guessed correctly it was: {number_to_guess}")
else:
    print(f"You have failed, the number was: {number_to_guess}")
```
:::
:::


## Write a Hangman game

```{pyodide}
#| exercise: hangman
DICTIONARY = ['hola', 'caracola', 'casa', 'barco']
INITIAL_NUM_ATTEMPS = 10

```

::: { .solution exercise="hangman" }
::: { .callout-tip collapse="false"}
```
import random

DICTIONARY = ['hola', 'caracola', 'casa', 'barco']
INITIAL_NUM_ATTEMPS = 10

def create_revealed_string(secret_word, guessed_letters):

    num_letters_missing = 0
    revealed_string = ''
    for letter in secret_word:
        if letter.upper() in guessed_letters:
            revealed_letter = letter
        else:
            revealed_letter = '-'
            num_letters_missing += 1
        revealed_string += revealed_letter
    return revealed_string


def play_game():
    secret_word = random.choice(DICTIONARY)
    guessed_letters = set()
    num_attempts = INITIAL_NUM_ATTEMPS
    won = None
    while True:
        guessed_letter = input('Pick a letter ')
        guessed_letters.add(guessed_letter.upper())

        revealed_string = create_revealed_string(secret_word, guessed_letters)
        num_attempts -= 1

        num_letters_to_guess = revealed_string.count('-')
        if not num_letters_to_guess:
            won = True
            break
        elif num_letters_to_guess:
            if num_attempts:
                print('The guess so far: ', revealed_string, guessed_letters)
            else:
                won = False
                break

    if won:
        print('Congratulations, you have won the game')
    else:
        print('You have lost')
    print('The secret word was: ', secret_word)


play_game()
```
:::
:::

## Read a fasta file

In bioinformatics we use fasta files to store DNA sequences.
This is an example of a fasta file with three sequences.

```
>seq1
CGCTAGCTAGTCTATCGATCTAGTCTAGCT
>seq2 some description after the space
TGTCGATCGTAGTCATCTGATCGACGTATCTA
CTCGAGTCATGCTATCATCATGCTAG
>seq3
TCAGTCGATGCTATCATCGTAGCTGATCGATCTGGCA
CTAGCAGTCGATC
```

Write a program that reads the sequences found in a fasta file and stores them in a list of dictionaries, one dictionary per sequence.

```{pyodide}
#| exercise: fasta
```

::: { .solution exercise="fasta" }
::: { .callout-tip collapse="false"}
```{pyodide}
from io import StringIO

def parse_fasta(fhand):
    seq = None
    for line in fhand:
        line = line.strip()
        if line[0] == ">":
            if seq:
                yield seq
            items = line[1:].split(" ", 1)
            if len(items) == 1:
                seq = {"name": items[0], "seq": ""}
            elif len(items) == 2:
                name, description = items
                seq = {"name": name, "seq": "", "description": description}
            else:
                raise RuntimeError("Fixme, we should not be here")
        else:
            seq["seq"] += line
    if seq:
        yield seq

if __name__ == "__main__":
    fasta_fhand = StringIO(">seq1 description\nATCT\nACTGC\n>seq2\nACTGTG\n")
    seqs = parse_fasta(fasta_fhand)
    print(list(seqs))
```
:::
:::

Now, create functions capable of calculating the percentage of GCs in the sequences and their lengths.

##

## Create a restriction map

Restriction enzymes recognize specific sequences, for instance:

- EcoRI: GATTC (cuts in the position between G and A, G^ATTC)
- Hind III : AAGCTT (cuts in the position between A and A, AAGCT^T)


Given a sequence, create the restriction fragments that these two sequences would create:

```{pyodide}
#| exercise: restriction_map
puc18_seq = """GCGCCCAATACGCAAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTT
CCCGACTGGAAAGCGGGCAGTGAGCGCAACGCAATTAATGTGAGTTAGCTCACTCATTAGGCACCCCAGG
CTTTACACTTTATGCTTCCGGCTCGTATGTTGTGTGGAATTGTGAGCGGATAACAATTTCACACAGGAAA
CAGCTATGACCATGATTACGAATTCGAGCTCGGTACCCGGGGATCCTCTAGAGTCGACCTGCAGGCATGC
AAGCTTGGCACTGGCCGTCGTTTTACAACGTCGTGACTGGGAAAACCCTGGCGTTACCCAACTTAATCGC
CTTGCAGCACATCCCCCTTTCGCCAGCTGGCGTAATAGCGAAGAGGCCCGCACCGATCGCCCTTCCCAAC
AGTTGCGCAGCCTGAATGGCGAATGGCGCCTGATGCGGTATTTTCTCCTTACGCATCTGTGCGGTATTTC
ACACCGCATATGGTGCACTCTCAGTACAATCTGCTCTGATGCCGCATAGTTAAGCCAGCCCCGACACCCG
CCAACACCCGCTGACGCGCCCTGACGGGCTTGTCTGCTCCCGGCATCCGCTTACAGACAAGCTGTGACCG
TCTCCGGGAGCTGCATGTGTCAGAGGTTTTCACCGTCATCACCGAAACGCGCGAGACGAAAGGGCCTCGT
GATACGCCTATTTTTATAGGTTAATGTCATGATAATAATGGTTTCTTAGACGTCAGGTGGCACTTTTCGG
GGAAATGTGCGCGGAACCCCTATTTGTTTATTTTTCTAAATACATTCAAATATGTATCCGCTCATGAGAC
AATAACCCTGATAAATGCTTCAATAATATTGAAAAAGGAAGAGTATGAGTATTCAACATTTCCGTGTCGC
CCTTATTCCCTTTTTTGCGGCATTTTGCCTTCCTGTTTTTGCTCACCCAGAAACGCTGGTGAAAGTAAAA
GATGCTGAAGATCAGTTGGGTGCACGAGTGGGTTACATCGAACTGGATCTCAACAGCGGTAAGATCCTTG
AGAGTTTTCGCCCCGAAGAACGTTTTCCAATGATGAGCACTTTTAAAGTTCTGCTATGTGGCGCGGTATT
ATCCCGTATTGACGCCGGGCAAGAGCAACTCGGTCGCCGCATACACTATTCTCAGAATGACTTGGTTGAG
TACTCACCAGTCACAGAAAAGCATCTTACGGATGGCATGACAGTAAGAGAATTATGCAGTGCTGCCATAA
CCATGAGTGATAACACTGCGGCCAACTTACTTCTGACAACGATCGGAGGACCGAAGGAGCTAACCGCTTT
TTTGCACAACATGGGGGATCATGTAACTCGCCTTGATCGTTGGGAACCGGAGCTGAATGAAGCCATACCA
AACGACGAGCGTGACACCACGATGCCTGTAGCAATGGCAACAACGTTGCGCAAACTATTAACTGGCGAAC
TACTTACTCTAGCTTCCCGGCAACAATTAATAGACTGGATGGAGGCGGATAAAGTTGCAGGACCACTTCT
GCGCTCGGCCCTTCCGGCTGGCTGGTTTATTGCTGATAAATCTGGAGCCGGTGAGCGTGGGTCTCGCGGT
ATCATTGCAGCACTGGGGCCAGATGGTAAGCCCTCCCGTATCGTAGTTATCTACACGACGGGGAGTCAGG
CAACTATGGATGAACGAAATAGACAGATCGCTGAGATAGGTGCCTCACTGATTAAGCATTGGTAACTGTC
AGACCAAGTTTACTCATATATACTTTAGATTGATTTAAAACTTCATTTTTAATTTAAAAGGATCTAGGTG
AAGATCCTTTTTGATAATCTCATGACCAAAATCCCTTAACGTGAGTTTTCGTTCCACTGAGCGTCAGACC
CCGTAGAAAAGATCAAAGGATCTTCTTGAGATCCTTTTTTTCTGCGCGTAATCTGCTGCTTGCAAACAAA
AAAACCACCGCTACCAGCGGTGGTTTGTTTGCCGGATCAAGAGCTACCAACTCTTTTTCCGAAGGTAACT
GGCTTCAGCAGAGCGCAGATACCAAATACTGTCCTTCTAGTGTAGCCGTAGTTAGGCCACCACTTCAAGA
ACTCTGTAGCACCGCCTACATACCTCGCTCTGCTAATCCTGTTACCAGTGGCTGCTGCCAGTGGCGATAA
GTCGTGTCTTACCGGGTTGGACTCAAGACGATAGTTACCGGATAAGGCGCAGCGGTCGGGCTGAACGGGG
GGTTCGTGCACACAGCCCAGCTTGGAGCGAACGACCTACACCGAACTGAGATACCTACAGCGTGAGCTAT
GAGAAAGCGCCACGCTTCCCGAAGGGAGAAAGGCGGACAGGTATCCGGTAAGCGGCAGGGTCGGAACAGG
AGAGCGCACGAGGGAGCTTCCAGGGGGAAACGCCTGGTATCTTTATAGTCCTGTCGGGTTTCGCCACCTC
TGACTTGAGCGTCGATTTTTGTGATGCTCGTCAGGGGGGCGGAGCCTATGGAAAAACGCCAGCAACGCGG
CCTTTTTACGGTTCCTGGCCTTTTGCTGGCCTTTTGCTCACATGTTCTTTCCTGCGTTATCCCCTGATTC
TGTGGATAACCGTATTACCGCCTTTGAGTGAGCTGATACCGCTCGCCGCAGCCGAACGACCGAGCGCAGC
GAGTCAGTGAGCGAGGAAGCGGAAGA"""

ecori = {"sequence": "GATTC", "cut-position": 1}
hindiii = {"sequence": "AAGCTT", "cut-position": 1}
enzymes = [ecori, hindiii]

print(cut_dna("GATTCAAGCTT", enzymes))
# expected fragments ['G', 'ATTCA', 'AGCTT']

puc18_seq = "".join(puc18_seq.splitlines())
fragments = cut_dna(puc18_seq, enzymes)
print(fragments)
```
::: { .solution exercise="restriction_map" }
::: { .callout-tip collapse="false"}
```{pyodide}
puc18_seq = """GCGCCCAATACGCAAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTT
CCCGACTGGAAAGCGGGCAGTGAGCGCAACGCAATTAATGTGAGTTAGCTCACTCATTAGGCACCCCAGG
CTTTACACTTTATGCTTCCGGCTCGTATGTTGTGTGGAATTGTGAGCGGATAACAATTTCACACAGGAAA
CAGCTATGACCATGATTACGAATTCGAGCTCGGTACCCGGGGATCCTCTAGAGTCGACCTGCAGGCATGC
AAGCTTGGCACTGGCCGTCGTTTTACAACGTCGTGACTGGGAAAACCCTGGCGTTACCCAACTTAATCGC
CTTGCAGCACATCCCCCTTTCGCCAGCTGGCGTAATAGCGAAGAGGCCCGCACCGATCGCCCTTCCCAAC
AGTTGCGCAGCCTGAATGGCGAATGGCGCCTGATGCGGTATTTTCTCCTTACGCATCTGTGCGGTATTTC
ACACCGCATATGGTGCACTCTCAGTACAATCTGCTCTGATGCCGCATAGTTAAGCCAGCCCCGACACCCG
CCAACACCCGCTGACGCGCCCTGACGGGCTTGTCTGCTCCCGGCATCCGCTTACAGACAAGCTGTGACCG
TCTCCGGGAGCTGCATGTGTCAGAGGTTTTCACCGTCATCACCGAAACGCGCGAGACGAAAGGGCCTCGT
GATACGCCTATTTTTATAGGTTAATGTCATGATAATAATGGTTTCTTAGACGTCAGGTGGCACTTTTCGG
GGAAATGTGCGCGGAACCCCTATTTGTTTATTTTTCTAAATACATTCAAATATGTATCCGCTCATGAGAC
AATAACCCTGATAAATGCTTCAATAATATTGAAAAAGGAAGAGTATGAGTATTCAACATTTCCGTGTCGC
CCTTATTCCCTTTTTTGCGGCATTTTGCCTTCCTGTTTTTGCTCACCCAGAAACGCTGGTGAAAGTAAAA
GATGCTGAAGATCAGTTGGGTGCACGAGTGGGTTACATCGAACTGGATCTCAACAGCGGTAAGATCCTTG
AGAGTTTTCGCCCCGAAGAACGTTTTCCAATGATGAGCACTTTTAAAGTTCTGCTATGTGGCGCGGTATT
ATCCCGTATTGACGCCGGGCAAGAGCAACTCGGTCGCCGCATACACTATTCTCAGAATGACTTGGTTGAG
TACTCACCAGTCACAGAAAAGCATCTTACGGATGGCATGACAGTAAGAGAATTATGCAGTGCTGCCATAA
CCATGAGTGATAACACTGCGGCCAACTTACTTCTGACAACGATCGGAGGACCGAAGGAGCTAACCGCTTT
TTTGCACAACATGGGGGATCATGTAACTCGCCTTGATCGTTGGGAACCGGAGCTGAATGAAGCCATACCA
AACGACGAGCGTGACACCACGATGCCTGTAGCAATGGCAACAACGTTGCGCAAACTATTAACTGGCGAAC
TACTTACTCTAGCTTCCCGGCAACAATTAATAGACTGGATGGAGGCGGATAAAGTTGCAGGACCACTTCT
GCGCTCGGCCCTTCCGGCTGGCTGGTTTATTGCTGATAAATCTGGAGCCGGTGAGCGTGGGTCTCGCGGT
ATCATTGCAGCACTGGGGCCAGATGGTAAGCCCTCCCGTATCGTAGTTATCTACACGACGGGGAGTCAGG
CAACTATGGATGAACGAAATAGACAGATCGCTGAGATAGGTGCCTCACTGATTAAGCATTGGTAACTGTC
AGACCAAGTTTACTCATATATACTTTAGATTGATTTAAAACTTCATTTTTAATTTAAAAGGATCTAGGTG
AAGATCCTTTTTGATAATCTCATGACCAAAATCCCTTAACGTGAGTTTTCGTTCCACTGAGCGTCAGACC
CCGTAGAAAAGATCAAAGGATCTTCTTGAGATCCTTTTTTTCTGCGCGTAATCTGCTGCTTGCAAACAAA
AAAACCACCGCTACCAGCGGTGGTTTGTTTGCCGGATCAAGAGCTACCAACTCTTTTTCCGAAGGTAACT
GGCTTCAGCAGAGCGCAGATACCAAATACTGTCCTTCTAGTGTAGCCGTAGTTAGGCCACCACTTCAAGA
ACTCTGTAGCACCGCCTACATACCTCGCTCTGCTAATCCTGTTACCAGTGGCTGCTGCCAGTGGCGATAA
GTCGTGTCTTACCGGGTTGGACTCAAGACGATAGTTACCGGATAAGGCGCAGCGGTCGGGCTGAACGGGG
GGTTCGTGCACACAGCCCAGCTTGGAGCGAACGACCTACACCGAACTGAGATACCTACAGCGTGAGCTAT
GAGAAAGCGCCACGCTTCCCGAAGGGAGAAAGGCGGACAGGTATCCGGTAAGCGGCAGGGTCGGAACAGG
AGAGCGCACGAGGGAGCTTCCAGGGGGAAACGCCTGGTATCTTTATAGTCCTGTCGGGTTTCGCCACCTC
TGACTTGAGCGTCGATTTTTGTGATGCTCGTCAGGGGGGCGGAGCCTATGGAAAAACGCCAGCAACGCGG
CCTTTTTACGGTTCCTGGCCTTTTGCTGGCCTTTTGCTCACATGTTCTTTCCTGCGTTATCCCCTGATTC
TGTGGATAACCGTATTACCGCCTTTGAGTGAGCTGATACCGCTCGCCGCAGCCGAACGACCGAGCGCAGC
GAGTCAGTGAGCGAGGAAGCGGAAGA"""


def cut_dna(dna_seq, enzymes):
    cut_positions = set()
    for enzyme in enzymes:
        enzyme_pos = dna_seq.find(enzyme["sequence"])
        if enzyme_pos < 0:
            continue
        cut_pos = enzyme_pos + enzyme["cut-position"]
        cut_positions.add(cut_pos)
    cut_positions = sorted(cut_positions)

    fragments = []
    for cut_position in reversed(cut_positions):
        fragment = dna_seq[cut_position:]
        dna_seq = dna_seq[:cut_position]
        fragments.append(fragment)
    if dna_seq:
        fragments.append(dna_seq)
    return list(reversed(fragments))

ecori = {"sequence": "GATTC", "cut-position": 1}
hindiii = {"sequence": "AAGCTT", "cut-position": 1}
enzymes = [ecori, hindiii]

print(cut_dna("GATTCAAGCTT", enzymes))
# expected fragments ['G', 'ATTCA', 'AGCTT']
puc18_seq = "".join(puc18_seq.splitlines())
fragments = cut_dna(puc18_seq, enzymes)
```
:::
:::
